src/liquidator.ts

```diff
@@ -0,0 +1,647 @@
+            await this.clearLiquidationGains();
+        }), (error => {
+            console.error(error);
+        }), this.liquidationsQueue);
+
+        // add the borrowing market state PublicKey to the polling accounts fetcher
+        this.pollingAccountsFetcher.addProgram('borrowingMarketState', clusterConfig.borrowing.accounts.borrowingMarketState, this.hubbleProgram, (async (borrowingMarketState : BorrowingMarketState) => {
+            this.borrowingMarketState = borrowingMarketState;
+
+            // will need to recalculate the minimum collateral ratio
+            this.mcr = await this.calculateMcr();
+        }), (error => {
+            console.error(error);
+        }), this.borrowingMarketState);
+    
+
+        // initiate the token prices object
+        this.prices = {} as TokenPrices;
+
+        // loop through the PriceInfo PublicKeys from the pyth object on the cluster config
+        Object.keys(this.clusterConfig.borrowing.accounts.pyth).filter(key => key.includes('PriceInfo')).forEach(key => {
+
+            // set prices to zero temporarily
+            this.prices[key.split('Price')[0]] = { value: new Decimal(0), exp: new Decimal(0) };
+            
+            // add the PublicKey of the pyth oracle to the polling accounts fetcher
+            // since the @pythnetwork/client doesn't have an anchor IDL which I can load into a program
+            // instead this `addConstructAccount` will parse the price data returned from the rpc using `parsePriceData`
+            this.pollingAccountsFetcher.addConstructAccount(this.clusterConfig.borrowing.accounts.pyth[key], (data: Buffer) => {
+                return parsePriceData(data);
+            }, async (priceData: PriceData) => {
+                // sometimes the price/exponent is undefined because of the pyth oracle
+                if (priceData.price && priceData.exponent) {
+                    this.prices[key.split('Price')[0]] = { value: new Decimal(priceData.price), exp: new Decimal(priceData.exponent) };
+                    // calculate the minimum collateral ratio whenever the price changes
+                    this.mcr = await this.calculateMcr();
+                } else {
+                    console.log(`Pyth Oracle: ${priceData.productAccountKey.toBase58()} - ${Object.keys(this.clusterConfig.borrowing.accounts.pyth).map(key => ({ key, value: this.clusterConfig.borrowing.accounts.pyth[key] } )).find(keyValue => keyValue.value === priceData.productAccountKey.toBase58()).key.split("Pr")[0].toUpperCase()} - price ${PriceStatus[priceData.status]}`);
+                }
+            }, (error) => {
+                console.error(error);
+            });
+        });
+    }
+
+    // create the clearLiquidationGains TransactionInstruction for the liquidator running this bot
+    // will need to be called for each token
+    async getClearLiquidationGainsIX(token: string, clearerAassociatedTokenAccount: PublicKey) : Promise<TransactionInstruction> {
```

Comment:
you could also generate an anchor Provider client that already has the rpc endpoints built in